vms/platformvm/state/state.go

```diff
@@ -121,11 +125,40 @@ type State interface {
 	// [vdrs].
 	ValidatorSet(subnetID ids.ID, vdrs validators.Set) error
 
-	GetValidatorWeightDiffs(height uint64, subnetID ids.ID) (map[ids.NodeID]*ValidatorWeightDiff, error)
-
-	// Returns a map of node ID --> BLS Public Key for all validators
-	// that left the Primary Network validator set.
-	GetValidatorPublicKeyDiffs(height uint64) (map[ids.NodeID]*bls.PublicKey, error)
+	// ApplyValidatorWeightDiffs iterates from [startHeight] towards the genesis
+	// block until it has applied all of the diffs through [endHeight]. Applying
+	// the diffs results in modifying [validators].
+	//
+	// Invariant: If attempting to generate the validator set for
+	// [endHeight - 1], [validators] should initially contain the validator
```

Comment:
should or must?

Reply:
It won't panic if this isn't the case... But it could error... I feel like `should` captures that...Technically, the user of this API could include additional zero values and the result would still work... (but that would be weird)... idk I'm not picky on the wording here... If you want it to say `must` then I'm fine changing it.

Reply:
I feel like we should just say "must" in the invariant -- better to document how we expect it to be used than how it _can_ (but probably shouldn't) be used.